Support Custom Secret Targets

ehazlett opened the pull request:
This adds support for custom secret target paths lifting the restriction of limiting secrets to `/var/run`.   Here is an example:

```
$> docker service create --name test --secret source=app,target=/etc/app -t alpine ash
# brace yourself for this one-liner...  yolo
$> docker exec -ti $(docker inspect --format '{{.Status.ContainerStatus.ContainerID}}' $(docker service ps --format '{{.ID}}' test)) mount | grep tmpfs | grep app
tmpfs on /etc/app type tmpfs (ro,relatime)
$> docker exec -ti $(docker inspect --format '{{.Status.ContainerStatus.ContainerID}}' $(docker service ps --format '{{.ID}}' test)) cat /etc/app
TESTING
```

Refs https://github.com/docker/docker/pull/32336#issuecomment-293035306
Depends on https://github.com/docker/swarmkit/pull/2118

ehazlett committed:
support custom paths for secrets

This adds support to specify custom container paths for secrets.

Signed-off-by: Evan Hazlett <ejhazlett@gmail.com>

ehazlett commented:
/cc @tianon @thaJeztah @diogomonica

aaronlehmann commented on vendor.conf:
```diff
@@ -105,7 +105,8 @@ github.com/docker/containerd 9048e5e50717ea4497b757314bad98ea3763c145
 # cluster
-github.com/docker/swarmkit d5232280c510d70755ab11305d46a5704735371a
+github.com/docker/swarmkit d58bab69b2305019be600726bca5f82392f4f116 https://github.com/ehazlett/swarmkit
+github.com/golang/mock bd3c8e81be01eef76d4b503f5e687d2d1354d2d9
```
Don't re-add mock. It was removed here: #32511

aaronlehmann commented on integration-cli/docker_cli_service_create_test.go:
```diff
@@ -90,35 +90,48 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
-	var refs []swarm.SecretReference
-	c.Assert(json.Unmarshal([]byte(out), &refs), checker.IsNil)
-	c.Assert(refs, checker.HasLen, 1)
+func (s *DockerSwarmSuite) TestServiceCreateWithSecretSourceTargetPaths(c *check.C) {
```
Slightly out of scope, but I realized recently that we don't seem to have tests that actually exec into the container to check that the secrets exist in the right place and have the right content. It's been on my list to add this. If you have the bandwidth to do it, I think it would be very valuable. Inspecting the container only goes so far - it doesn't confirm the secret mounts are working as expected.

ehazlett replied:
Same here.  I'll get some added.

aaronlehmann commented on container/container_unix.go:
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
+		srcPath := filepath.Join(container.SecretMountPath(), filepath.Base(r.File.Name))
+		mounts = append(mounts, Mount{
+			Source:      srcPath,
+			Destination: r.File.Name,
```
Doesn't this need to be prefixed with `containerSecretMountPath` if the target is not an absolute path?

aaronlehmann commented on container/container_unix.go:
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
+	for _, r := range container.SecretReferences {
+		// secrets are created in the SecretMountPath at a single level
+		// i.e. /var/run/secrets/foo
+		srcPath := filepath.Join(container.SecretMountPath(), filepath.Base(r.File.Name))
```
What happens if two secret targets have the same base name?

aaronlehmann commented on container/container_unix.go:
```diff
@@ -253,29 +253,40 @@ func (container *Container) IpcMounts() []Mount {
-		if os.IsNotExist(err) {
-			return nil
+	for _, r := range container.SecretReferences {
+		srcPath := filepath.Join(container.SecretMountPath(), r.File.Name)
```
In `SecretMounts` this uses the base name. The code here doesn't seem consistent.

Either way, we should break this out into a shared function.

ehazlett replied:
Good point :+1:

cpuguy83 commented on integration-cli/docker_cli_service_create_test.go:
```diff
@@ -90,35 +90,48 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
-	c.Assert(refs[0].File.Name, checker.Equals, testTarget)
+	testPaths := map[string]string{
+		"app":         "/etc/secret",
+		"test_secret": "test_secret",
```
Can we test directory traversal?

thaJeztah commented:
Design looks good to me 👍 Moving this to code review, but let me know if there's objections 😄

aaronlehmann commented:
It still looks like this will break when two secrets have the same target base name.

aaronlehmann commented:
@ehazlett: I'd suggest this borrow the code from configs in #32336. That stores the config under the full target path, so there isn't a risk of collisions from targets that share the same basename. Also, it would be good for the config and secret implementations to stay consistent.

Let me know if it would be helpful for me to carry this PR.

ehazlett commented:
Ya if you want to carry that's fine with me.

Commit:
Use "local" secret paths based on the secretID

This prevents targets with the same basename from colliding.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Extend test coverage of secrets

Actually look inside the container to see if the secret data is present
and correct. Test absolute paths, relative paths, and just a basename.
Test the scenario where a service references the same secret under
different targets.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
I've updated the vendoring, as the upstream swarmkit PR was merged.

I added test coverage for the following:

- Verify the presence and contents of secrets with simple, relative, and absolute targets
- Ensure that directory traversal is prevented
- Check that a service can reference the same secret multiple times under different targets

I also added corresponding tests to the configs PR (#32336).

PTAL

thaJeztah commented:
Thanks @aaronlehmann!

ping @cyli @riyazdf ptal

aaronlehmann commented:
`TestSwarmLockUnlockCluster` appears to have hung on `powerpc` and `z`. I'd guess this is related to the swarmkit vendoring rather than the specific changes in this PR. We're working on getting an etcd/raft issue fixed upstream - this might be related.

diogomonica commented on container/container.go:
```diff
@@ -949,3 +949,26 @@ func (container *Container) InitializeStdio(iop libcontainerd.IOPipe) error {
+// SecretFilePath returns the path to the location of a secret on the host.
+func (container *Container) SecretFilePath(secretRefTarget *swarmtypes.SecretReferenceFileTarget) (string, error) {
+	secretMountPath := container.SecretMountPath()
+	mountPath := filepath.Join(secretMountPath, secretRefTarget.Name)
```
Are we ok with joining `secretRefTarget.Name` directly? I'm assuming our regex doesn't allow any path separator characters in names? (it definitely allows `.`)

aaronlehmann replied:
It does allow path separators - that's the point of https://github.com/docker/swarmkit/pull/2118.

The check below tries to ensure that this doesn't escape the mount point, but I'd appreciate careful review to make sure this works as intended.

diogomonica commented on container/container_unix.go:
```diff
@@ -253,29 +248,52 @@ func (container *Container) IpcMounts() []Mount {
-			Source:      container.SecretMountPath(),
-			Destination: containerSecretMountPath,
-			Writable:    false,
+// SecretMounts returns the mount for the secret path
```
*mounts

diogomonica commented on integration-cli/docker_cli_service_create_test.go:
```diff
@@ -90,35 +91,128 @@ func (s *DockerSwarmSuite) TestServiceCreateWithSecretSimple(c *check.C) {
+		"app":             "/etc/secret",
+		"test_secret":     "test_secret",
+		"relative_secret": "relative/secret",
+		"invalid_secret":  "../secret",
```
Should we even allow a secret create w/ a target like this?

aaronlehmann replied:
The way paths work is OS-specific, so deciding in the manager or the client whether a path is acceptable is a bit tricky. We don't know what OS the service will eventually run on.

cyli replied:
Would checking `path == filepath.Clean(path)` work across platform, and just clean any references to `..`?  It might be more explicit to fail this way.

If I understand correctly, this particular case is invalid because on the host it would take you out of the flat mount directory, but something like `ignoreddir/../test_secret` would be fine because that would end up being just `secret`.  That might be confusing to users, since they may just want it in `/var/run/test_secret` as opposed to `/var/run/secrets/test_secret`, and may not understand that the `../` is disallowed because it's taking them out of the host path specifically.

aaronlehmann replied:
I don't think it would work across platforms, because there are different implementations for Windows and Linux. Even if it does work, I wouldn't be super comfortable with passing Windows paths to `filepath.Clean` on a Linux host and vice versa.

I suppose we could grep for `(^|/)\.\./` and `(^|\)\.\.\`. This would take a PR to swarmkit, though.

cyli replied:
Would it be feasible to have the filename be the secret ID or something else that wouldn't be traversable in the host directory?  That way we wouldn't have to be afraid of traversals out of the container root.

cpuguy83 replied:
We can at least let the worker evaluate the path and reject it. Also see: https://github.com/moby/moby/blob/master/pkg/symlink/fs.go#L21

diogomonica commented on opts/secret.go:
```diff
@@ -53,10 +52,6 @@ func (o *SecretOpt) Set(value string) error {
 		case "source", "src":
 			options.SecretName = value
 		case "target":
-			tDir, _ := filepath.Split(value)
```
Why are we fine with a target being a directory? This still doesn't make sense right?

aaronlehmann replied:
The client code has no way of knowing whether a path is a file or a directory.

This old code was essentially checking that there was no path separator. It is being removed because now path separators are allowed.

diogomonica reviewed (commented):
Couple of questions:
- Should we even support secret creation with targets that have potential OS separators?
- Should we list all secrets in `/var/run/secrets` even if the target is outside of base secrets dir? Would have the advantage that a `ls /var/run/secrets/` always returns all the secrets available.
- Should we allow someone to add a target into `/var/run/secrets/`? In particular concerned that the validation of multiple targets being set somehow doesn't handle this new way of adding targets and the behavior will be unknown or the task unschedulable.
- Are there any files/paths we want to reject outright from being overwritten? (like anything in `/proc/`?)

Thanks for carrying this PR @aaronlehmann

aaronlehmann commented:
> Should we even support secret creation with targets that have potential OS separators?

This one I don't understand. The purpose of this PR is to allow you to put secrets anywhere on the filesystem, not just `/run/secrets`. For this to work, it must be possible to include path separators in the target. Do you mean something different by "OS separators"?

> Should we list all secrets in /var/run/secrets even if the target is outside of base secrets dir? Would have the advantage that a ls /var/run/secrets/ always returns all the secrets available.

I'm not sure how it would work. What would we use as the filename under `/run/secrets` that would be guaranteed to be unique? I suppose we could use secret ID but it would not be very user-friendly. Or we could put a secret with the target `/path/to/target` in both `/path/to/target` and `/run/secrets/path/to/target`, but I feel like that might be confusing.

> Should we allow someone to add a target into /var/run/secrets/? In particular concerned that the validation of multiple targets being set somehow doesn't handle this new way of adding targets and the behavior will be unknown or the task unschedulable.

Let's make sure no two secrets have the same `getSecretTargetPath` result.

I'm okay with also disallowing targets that explicitly use a path under `/run/secrets`, but the above would be sufficient to prevent conflicts.

> Are there any files/paths we want to reject outright from being overwritten? (like anything in /proc/?)

Since we use bind mounts, I don't think we're actually overwriting anything. I can't think of any paths that would be important to block, but let me know if you think of any.

aaronlehmann commented:
@diogomonica: I checked, and putting a target in `/run/secrets` in such a way that it conflicts with an implicit path is already prevented:

```
ERRO[0125] fatal task error                              error="starting container failed: linux mounts: Duplicate mount point '/run/secrets/foo'" module="node/agent/taskmanager" node.id=eztzkyt8y44569gywcgl7t03a service.id=02usncjhurcy4di7q5wrv3ij2 task.id=qvrp1dwru2ittc7dhg560xluy
```

diogomonica commented:
LGTM

cyli commented on container/container_windows.go:
```diff
@@ -43,8 +47,8 @@ func (container *Container) IpcMounts() []Mount {
 
-// SecretMount returns the mount for the secret path
-func (container *Container) SecretMount() *Mount {
+// SecretMounts returns the mount for the secret path
```
nitpick:  "mounts" here too :)

cyli commented on daemon/container_operations_unix.go:
```diff
@@ -177,13 +177,15 @@ func (daemon *Daemon) setupSecretDir(c *container.Container) (setupErr error) {
 
-		fPath := filepath.Join(localMountPath, targetPath)
+		// secrets are created in the SecretMountPath at a single level
+		// i.e. /var/run/secrets/foo
```
Non-blocking comment nit:  Should this be something like `/var/lib/docker/devicemapper/mnt/<container sha>/secrets/foo` (even though they may not be using devicemapper, just to make it clear that `SecretMountPath` is on the host and not in a container?

aaronlehmann commented:
Fixed nits

aaronlehmann commented:
I've pushed an update that uses the secret ID instead of the secret target as the filename for storing the secret on the host's tmpfs mount.

cyli commented:
Thanks @aaronlehmann!  I could not think of any issues allowing a secret to escape `/run/secrets` inside a container, since we already allow absolute paths inside the container, whomever provides the target can get to any path anyway.  

I don't think users should be providing traversals from `/run/secrets` rather than absolute paths, but other than them confusing themselves, I don't think it presents any vulnerability if we're just using IDs (which we control) on the host.

Thinking about it more, it may be nice if we could prevent path traversals at all on `service create` but since as you mentioned that could be tricky, I'm not sure it's worth the effort, but would appreciate others giving this line of reasoning a sanity check :)

LGTM on green.

diogomonica commented:
@cyli @aaronlehmann since we don't see any security issues and this would just be a hygienic decision (no transversal), it would probably just be best effort.

I also agree that we shouldn't allow relative paths at all. Unless someone can come up with a good use case.

aaronlehmann commented:
Rebased

Commit:
Use forked version of CLI for tests

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Corresponding CLI PR in https://github.com/docker/cli/pull/50 (really just vendoring `opts`, which is a bit odd...)

johnstep reviewed (approved):
LGTM

cpuguy83 commented:
```
Handler for DELETE /v1.30/containers/9e returned error: Unable to remove filesystem for 9e6c79a949b080473b9d4778f6eb9b3c9fb06a142260912457d57306a3e6d476: remove /var/lib/docker/containers/9e6c79a949b080473b9d4778f6eb9b3c9fb06a142260912457d57306a3e6d476/secrets: device or resource busy
```

This seems to happen with any secret target right now.

And I end up with leaked mounts:
```
tmpfs on /var/lib/docker/containers/67c604df32844744fea5d2613dc850d1c85227fc0a9406f4008938b4975358e4/secrets type tmpfs (rw,nosuid,nodev,noexec,relatime)
```

cpuguy83 commented on container/container_unix.go:
```diff
@@ -253,29 +248,43 @@ func (container *Container) IpcMounts() []Mount {
+			return err
 		}
-		return err
 	}
```
I think after this is done we need to `detachMounted(container.SecretMountPath())`

aaronlehmann replied:
You are right. I'm not sure why this got removed. I think we should `defer` the original contents of this function - do you agree?

cpuguy83 replied:
Yep, makes sense.

aaronlehmann replied:
@cpuguy83: I believe I've fixed this. I'm not seeing any errors or leaked mounts when I delete a container that uses secrets. I'd appreciate it if you could confirm on your side.

cpuguy83 commented on container/container_unix.go:
```diff
@@ -253,29 +248,59 @@ func (container *Container) IpcMounts() []Mount {
-	if _, err := os.Stat(container.SecretMountPath()); err != nil {
-		if os.IsNotExist(err) {
-			return nil
+func (container *Container) UnmountSecrets() (err error) {
```
Somewhere in here we are now getting a spurious EINVAL.
Maybe we can just check `if err == unix.EINVAL { if mount.Mounted(path) { return err }}`

Although I'm really curious as to where it's happening.

cpuguy83 replied:
I don't think we should return errors anywhere in this function since it stops all processing but cleanup should still happen for any remaining items.
The parent is only logging this error anyway.

Maybe if we do want to continue to return errors we can just collect all the errors into a single error.

ehazlett replied:
For a little more context:
```
DEBU[3206] containerd: process exited                    id=2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d pid=init status=0 systemPid=1631
DEBU[3206] libcontainerd: received containerd event: &types.Event{Type:"exit", Id:"2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d", Status:0x0, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc421526db0)} 
DEBU[3206] Revoking external connectivity on endpoint test.1.3g1qqhu9spa9apbz7t28ixkvv (a8d45de538bfa41fe4b92dc866eda1426a653ebf79e68480ee768905c96f764d) 
DEBU[3206] DeleteConntrackEntries purged ipv4:0, ipv6:0 
DEBU[3206] Releasing addresses for endpoint test.1.3g1qqhu9spa9apbz7t28ixkvv's interface on network bridge 
DEBU[3206] ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2) 
WARN[3206] 2759b664f3a609398c03bcbc6af6ee3e9996097c2363d63a7c5a0100cf79037d cleanup: failed to unmount secrets: invalid argument 
```

cpuguy83 commented on container/container_unix.go:
```diff
@@ -253,29 +248,59 @@ func (container *Container) IpcMounts() []Mount {
+		srcPath := container.SecretFilePath(*r)
+		if _, err := os.Stat(srcPath); err != nil {
+			if os.IsNotExist(err) {
+				return nil
```
continue

aaronlehmann commented:
I've reverted all changes to `UnmountSecrets`. I think these were incorrect to begin with. The individual secret mounts are passed on to `runc` and should be cleaned up by `runc`. All that's necessary is unmounting the tmpfs on the host.

PTAL

cpuguy83 commented on hack/dockerfile/binaries-commits:
```diff
@@ -11,5 +11,5 @@ VNDR_COMMIT=c56e082291115e369f77601f9c071dd0b87c7120
-DOCKERCLI_REPO=https://github.com/docker/cli
-DOCKERCLI_COMMIT=c3648a9c9400d45524cc71b8fca4085b192c626f
+DOCKERCLI_REPO=https://github.com/aaronlehmann/cli
+DOCKERCLI_COMMIT=fd5a5910409fe5ed862b29de45a66f2bf8056894
```
Need to change these back before merge.

aaronlehmann replied:
They can't be changed back before the merge. This would cause CI to fail - unless you think the CLI PR should be merged before this one (but that seems backwards).

These circular dependencies are very hard to deal with.

johnstep replied:
Don't we need to keep this until docker/cli#50 is merged?

cpuguy83 replied:
We can move the tests to a separate PR.

Opts package should really not be in this repo.
We're only using a couple of functions from it for cmd/dockerd.

aaronlehmann replied:
The plan I discussed with @tiborvass and @thaJeztah was to merge this PR with the `DOCKERCLI_REPO` change intact, then change this back to the main repo in a followup.

cpuguy83 replied:
Really even the opts change could be made on it's own since the existing code is not valid as is.

cpuguy83 replied:
@aaronlehmann That's fine.

cpuguy83 reviewed (approved):
LGTM

ehazlett commented:
FWIW tested on my end again and LGTM
